hazelcast/src/main/java/com/hazelcast/partition/strategy/AttributePartitioningStrategy.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.partition.strategy;
+
+import com.hazelcast.core.HazelcastJsonValue;
+import com.hazelcast.internal.serialization.SerializableByConvention;
+import com.hazelcast.internal.serialization.impl.GenericRecordQueryReader;
+import com.hazelcast.internal.serialization.impl.InternalGenericRecord;
+import com.hazelcast.jet.impl.util.ReflectionUtils;
+import com.hazelcast.nio.serialization.DataSerializable;
+import com.hazelcast.partition.PartitioningStrategy;
+import com.hazelcast.query.impl.getters.JsonGetter;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+@SerializableByConvention
+public class AttributePartitioningStrategy implements PartitioningStrategy {
+
+    private String[] attributes;
+
+    public AttributePartitioningStrategy(String... attributes) {
+        this.attributes = attributes;
+    }
+
+    @Override
+    public Object getPartitionKey(final Object key) {
+        final Object[] values = new Object[attributes.length];
+        if (key instanceof InternalGenericRecord) {
```

Comment:
`DefaultPartitioningStrategy` supports `PartitionAware`. Should we also support `PartitionAware` here (as first case maybe)?

Reply:
I meant that if `key` implements `PartitionAware` we could use this interface instead of extracting attributes. Current `AttributePartitioningStrategy` implementation ignores `PartitionAware`.

That's possibly not a best design of user's app if they store different key types in one IMap, but we support it.

Reply:
I think that would be a discrepancy from the behavior of other strategies. Yes it would be convenient(er) for end user, but every other strategy just falls back to basic logic rather than essentially try to fall back to GlobalStrategy, so for consistency sake alone its probably best to keep it as is. TL;DR: its convenient, but its also unexpected. 

FWIW IMO it would be better to throw an exception in this case (and other cases of incompatible entities being passed to the strategy). If someone configures Attribute strategy for their map, it would be ideal to alert them about a problem rather than just let the partitioning logic "eat" this issue.

Reply:
We must not do this. The point of declarative partitioning strategy is that the query engine knows by which attribute the keys are partitioned, and must not fall back to some opaque java code.